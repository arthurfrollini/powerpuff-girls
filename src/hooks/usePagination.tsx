import React, { createContext, ReactNode, useContext, useState } from "react";

import { Show } from "../infra/models/TvShow";

interface PaginationContextData {
  startIndex: number;
  pages: number;
  currentPage: number;
  setItems: React.Dispatch<React.SetStateAction<Show[]>>;
  setCurrentPage: React.Dispatch<React.SetStateAction<number>>;
  currentItems: Show[];
}

interface PaginationProps {
  children: ReactNode;
}

export const PaginationContext = createContext<PaginationContextData>(
  {} as PaginationContextData
);

export function PaginationProvider({ children }: PaginationProps) {
  const [items, setItems] = useState<Show[]>([]);
  const [itemsPerPage] = useState(10);
  const [currentPage, setCurrentPage] = useState(0);

  const startIndex = currentPage * itemsPerPage;
  const endIndex = startIndex + itemsPerPage;
  const currentItems = items.slice(startIndex, endIndex);

  const pages = Math.ceil(items.length / itemsPerPage);

  return (
    <PaginationContext.Provider
      value={{
        startIndex,
        pages,
        setItems,
        currentItems,
        setCurrentPage,
        currentPage,
      }}
    >
      {children}
    </PaginationContext.Provider>
  );
}

export const usePagination = () => useContext(PaginationContext);
