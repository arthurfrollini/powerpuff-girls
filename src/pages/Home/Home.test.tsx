import { render, screen } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";
import { PaginationProvider } from "../../hooks/usePagination";

import nock from "nock";
import axios from "axios";

import renderer from "react-test-renderer";
import { Home } from ".";

axios.defaults.adapter = require("axios/lib/adapters/http");

describe("Home component", () => {
  it("renders correctly", () => {
    render(
      <Router>
        <PaginationProvider>
          <Home />
        </PaginationProvider>
      </Router>
    );

    expect(screen.getByText("Lista de episÃ³dios:")).toBeInTheDocument();
  });

  it("fetch data correctly", async () => {
    const scope = nock("http://api.tvmaze.com")
      .get("/search/shows?q=powerpuff")
      .reply(200, "test response");

    await axios.get("http://api.tvmaze.com/search/shows?q=powerpuff");

    scope.done();
  });

  it("matches snapshot", () => {
    const tree = renderer.create(
      <Router>
        <PaginationProvider>
          <Home />
        </PaginationProvider>
      </Router>
    );

    expect(tree).toMatchSnapshot();
  });
});
