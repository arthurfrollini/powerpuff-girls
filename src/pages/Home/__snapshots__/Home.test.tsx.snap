// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home component matches snapshot 1`] = `
<div
  className="sc-bdvvtL gzBWRV"
>
  <strong
    className="sc-gsDKAQ yoBRS"
  >
    Lista de epis√≥dios:
  </strong>
  <div
    className="sc-kDTinF TzqHu"
  />
  <div
    className="sc-bqiRlB gtwUVt"
    data-testid="pagination"
  />
</div>
`;
